[libpas][PGM] PGM allocations should respect minimum alignment

dhecht opened the pull request:
#### f73ebb626eaa3d3426f082e65aadfdc560844b76
<pre>
[libpas][PGM] PGM allocations should respect minimum alignment
<a href="https://bugs.webkit.org/show_bug.cgi?id=286429">https://bugs.webkit.org/show_bug.cgi?id=286429</a>
<a href="https://rdar.apple.com/143439034">rdar://143439034</a>

Reviewed by Keith Miller.

PGM allocations need to be aligned. When &apos;right_align==false&apos;, allocations
are aligned to the page boundary, so there is no problem in that case.

However, when &apos;right_align==true&apos;, the returned allocation starts at
&apos;size&apos; bytes prior to a page boundary. When &apos;size&apos; is some unaligned
amount, this causes the allocation to be unaligned.

Various places in JSC assume the alignment of all byte buffers regardless
of their sizes (which is a valid assumption since malloc memory must be aligned
for any type) in order to use the low bits to store metadata. So this bug causes
various issues in JSC when one of these odd sized byte buffers is allocated
in PGM mode.

* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_allocate):
* Source/bmalloc/libpas/src/test/PGMTests.cpp:
(std::checkMalloc):
(std::testPGMSingleAlloc):
(std::testPGMMultipleAlloc):
(std::testPGMRealloc):
(std::testPGMMetaData):
(std::testPGMErrors):
(std::testPGMMetadataVectorManagement):
(std::testPGMMetadataVectorManagementFewDeallocations):
(std::testPGMMetadataDoubleFreeBehavior):
(std::testPGMMetadataVectorManagementRehash):

Canonical link: <a href="https://commits.webkit.org/289327@main">https://commits.webkit.org/289327@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0cb87d6261796388f660066f3f33845c19e46984

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86441 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5984 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40775 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91343 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37231 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6237 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14022 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66917 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24703 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89447 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4729 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78284 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47237 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4537 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32609 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36342 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/79280 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75057 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33484 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93222 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/85269 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13636 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9883 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/75721 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13843 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74151 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74904 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18431 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19161 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17548 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/6439 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13659 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/107735 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13411 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25929 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16855 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15195 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -85,6 +88,12 @@ pas_allocation_result pas_probabilistic_guard_malloc_allocate(pas_large_heap* la
 
     const size_t page_size = pas_page_malloc_alignment();
 
+    type = pas_heap_for_large_heap(large_heap)->type;
+    alignment = heap_config->get_type_alignment(type);
+    alignment = PAS_MAX(alignment, heap_config->large_alignment);
+
+    size = pas_round_up_to_power_of_2(size, alignment);
```
Are we guaranteed that `alignment` is a power of 2 here?

kmiller68 replied:
Generally, alignments have to be a power of 2. Additionally, you'd crash in a Debug build if it wasn't a power of two for some `heap_config`.

dhecht replied:
PAS_ASSERT is on even on release builds. But yes, it is an invariant that pas alignment fields are powers of 2.

kmiller68 reviewed (approved):
r=me.

dhecht committed:
[libpas][PGM] PGM allocations should respect minimum alignment
https://bugs.webkit.org/show_bug.cgi?id=286429
rdar://143439034

Reviewed by Keith Miller.

PGM allocations need to be aligned. When 'right_align==false', allocations
are aligned to the page boundary, so there is no problem in that case.

However, when 'right_align==true', the returned allocation starts at
'size' bytes prior to a page boundary. When 'size' is some unaligned
amount, this causes the allocation to be unaligned.

Various places in JSC assume the alignment of all byte buffers regardless
of their sizes (which is a valid assumption since malloc memory must be aligned
for any type) in order to use the low bits to store metadata. So this bug causes
various issues in JSC when one of these odd sized byte buffers is allocated
in PGM mode.

* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_allocate):
* Source/bmalloc/libpas/src/test/PGMTests.cpp:
(std::checkMalloc):
(std::testPGMSingleAlloc):
(std::testPGMMultipleAlloc):
(std::testPGMRealloc):
(std::testPGMMetaData):
(std::testPGMErrors):
(std::testPGMMetadataVectorManagement):
(std::testPGMMetadataVectorManagementFewDeallocations):
(std::testPGMMetadataDoubleFreeBehavior):
(std::testPGMMetadataVectorManagementRehash):

Canonical link: https://commits.webkit.org/289327@main
